src/imitation/algorithms/preference_comparisons.py

```diff
@@ -809,27 +784,192 @@ def _probability(self, rews1: th.Tensor, rews2: th.Tensor) -> th.Tensor:
+                    self.logger.record(metric, output["metrics"][metric])
+
+
+class RewardEnsembleTrainer(BasicRewardTrainer):
+    """Train a reward ensemble."""
+
+    _model: reward_nets.RewardEnsemble
+
+    def __init__(
+        self,
+        model: reward_nets.RewardEnsemble,
+        loss: RewardLoss,
+        batch_size: int = 32,
+        epochs: int = 1,
+        lr: float = 1e-3,
+        weight_decay: float = 0.0,
+        custom_logger: Optional[imit_logger.HierarchicalLogger] = None,
+    ):
+        """Create an ensemble trainer."""
+        super().__init__(
+            model,
+            loss,
+            batch_size,
+            epochs,
+            lr,
+            weight_decay,
+            custom_logger,
+        )
+
+    def _train(self, dataset: PreferenceDataset, epoch_multiplier: float = 1.0):
+        """Trains for `epoch_multiplier * self.epochs` epochs over `dataset`."""
+        dataloader = self._make_data_loader(dataset)
         epochs = round(self.epochs * epoch_multiplier)
 
         for _ in tqdm(range(epochs), desc="Training reward model"):
             for fragment_pairs, preferences in dataloader:
                 self.optim.zero_grad()
-                loss = self._loss(fragment_pairs, preferences)
+                losses = []
+                metrics = []
+                for member in self._model.members:
+                    output = self.loss(member, fragment_pairs, preferences)
+                    losses.append(output["loss"])
+                    metrics.append(output["metrics"])
+                losses = th.stack(losses)
+                loss = losses.mean()
                 loss.backward()
                 self.optim.step()
+
+                # Note here we are return all the losses not just the mean
+                # This will give us a histogram
                 self.logger.record("loss", loss.item())
+                self.logger.record("loss_std", loss.std().item())
+                self.logger.record("dist_loss", losses.detach().cpu().numpy())
+
+                # Turn metrics from a list of dictionaries into a dictionary of
+                # tensors. Again this should give us a histogram in tensorboard.
+                metrics = {k: th.stack([di[k] for di in metrics]) for k in metrics[0]}
+                for metric in metrics:
+                    self.logger.record(metric, metrics[metric].mean().item())
+                    self.logger.record(f"dist_{metric}", metrics[metric].cpu().numpy())
+
+
+def make_reward_trainer(
+    reward_model: reward_nets.RewardNet,
+    loss: RewardLoss,
+    reward_trainer_kwargs: Mapping[str, Any] = {},
+) -> RewardTrainer:
+    """Construct the correct type of reward trainer for this reward function."""
+    base_model = reward_model
+    while hasattr(base_model, "base"):
+        base_model = base_model.base
+
+    if isinstance(base_model, reward_nets.RewardEnsemble):
+        return RewardEnsembleTrainer(
+            base_model,
```

Comment:
It's intentional. The only case where this would not be good is if for some reason we wanted to apply shaping on top of an ensemble. The current trainer does not support this, and I do not see a use case for it.

Reply:
The issue is that wrappers will hide attributes list `ensemble.members` and `predict_processed_all`. Which we need access to train the ensemble.

Reply:
What's the use case where we want to have a wrapper around `RewardEnsemble` but ignore it during training? Is it just normalization, that's not used during training anyway and just for RL training?

Reply:
We might want to add a `AddSTDRewardWrapper` during training to encourage optimism. In fact the current default configuration will always do this but just leave alpha=0.

Reply:
If we don't want to support any wrapping at all I would advocate for removing AddSTDRewardWrapper entirely and moving the ability to add the standard deviation to the abstract class RewardNetWithVariance.

Reply:
OK, the `AddSTDRewardWrapper` use case makes sense for me (add it to RL training but not reward model training). My suggestion is that we have a whitelist of "compatible" wrappers, and strip off those for reward model training. Anything else we throw an error or issue a warning for.

Generally the heuristic here is it's worth minor inconvenience for users/adding new features (i.e. update the whitelist) to avoid subtle bugs (a wrapper getting stripped that people didn't expect).

This whole thing is a good example of a use case we should try to support more cleanly post-refactor.